util/predicates/cluster_predicates.go

```diff
@@ -43,3 +48,82 @@ func processIfClusterNotImported(logger logr.Logger, obj client.Object) bool {
 
 	return true
 }
+
+// ClusterWithReadyControlPlane returns a predicate that returns true only if the provided resource is a cluster with a
+// ready control plane.
+func ClusterWithReadyControlPlane(logger logr.Logger) predicate.Funcs {
+	return predicate.Funcs{
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			return processIfClusterReadyControlPlane(logger.WithValues("predicate", "ClusterWithReadyControlPlane", "eventType", "update"), e.ObjectNew)
+		},
+		CreateFunc: func(e event.CreateEvent) bool {
+			return processIfClusterReadyControlPlane(logger.WithValues("predicate", "ClusterWithReadyControlPlane", "eventType", "create"), e.Object)
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			return processIfClusterReadyControlPlane(logger.WithValues("predicate", "ClusterWithReadyControlPlane", "eventType", "delete"), e.Object)
+		},
+		GenericFunc: func(e event.GenericEvent) bool {
+			return processIfClusterReadyControlPlane(logger.WithValues("predicate", "ClusterWithReadyControlPlane", "eventType", "generic"), e.Object)
+		},
+	}
+}
+
+func processIfClusterReadyControlPlane(logger logr.Logger, obj client.Object) bool {
+	kind := strings.ToLower(obj.GetObjectKind().GroupVersionKind().Kind)
+	log := logger.WithValues("namespace", obj.GetNamespace(), kind, obj.GetName())
+
+	cluster, ok := obj.(*clusterv1.Cluster)
+	if !ok {
+		panic(fmt.Sprintf("Expected a Cluster but got a %T", obj)) // it should never happen
```

Comment:
Don’t panic here. Return false, this just means that some other object came from cache, which could potentially happen.

Reply:
I agree, maybe log an error and return false?